lncfg/db.go

```diff
@@ -79,54 +107,267 @@ func (db *DB) Init(ctx context.Context, dbPath string) error {
-	RemoteDB kvdb.Backend
+	// MacaroonDB points to a database backend that stores the macaroon root
+	// keys.
+	MacaroonDB kvdb.Backend
+
+	// DecayedLogDB points to a database backend that stores the decayed log
+	// data.
+	DecayedLogDB kvdb.Backend
+
+	// TowerClientDB points to a database backend that stores the watchtower
+	// client data. This might be nil if the watchtower client is disabled.
+	TowerClientDB kvdb.Backend
+
+	// TowerServerDB points to a database backend that stores the watchtower
+	// server data. This might be nil if the watchtower server is disabled.
+	TowerServerDB kvdb.Backend
+
+	// WalletDB is an option that instructs the wallet loader where to load
+	// the underlying wallet database from.
+	WalletDB btcwallet.LoaderOption
+
+	// Remote indicates whether the database backends are remote, possibly
+	// replicated instances or local bbolt backed databases.
+	Remote bool
+
+	// CloseFuncs is a map of close functions for each of the initialized
+	// DB backends keyed by their namespace name.
+	CloseFuncs map[string]func() error
 }
 
-// GetBackends returns a set of kvdb.Backends as set in the DB config.  The
-// local database will ALWAYS be non-nil, while the remote database will only
-// be populated if etcd is specified.
-func (db *DB) GetBackends(ctx context.Context, dbPath string) (
-	*DatabaseBackends, error) {
+// GetBackends returns a set of kvdb.Backends as set in the DB config.
+func (db *DB) GetBackends(ctx context.Context, chanDBPath,
+	walletDBPath, towerServerDBPath string, towerClientEnabled,
+	towerServerEnabled bool) (*DatabaseBackends, error) {
 
-	var (
-		localDB, remoteDB kvdb.Backend
-		err               error
-	)
+	// We keep track of all the kvdb backends we actually open and return a
+	// reference to their close function so they can be cleaned up properly
+	// on error or shutdown.
+	closeFuncs := make(map[string]func() error)
+
+	// If we need to return early because of an error, we invoke any close
+	// function that has been initialized so far.
+	returnEarly := true
+	defer func() {
+		if !returnEarly {
+			return
+		}
+
+		for _, closeFunc := range closeFuncs {
+			_ = closeFunc()
+		}
+	}()
 
 	if db.Backend == EtcdBackend {
-		remoteDB, err = kvdb.Open(
-			kvdb.EtcdBackendName, ctx, db.Etcd,
+		// As long as the graph data, channel state and height hint
+		// cache are all still in the channel.db file in bolt, we
+		// replicate the same behavior here and use the same etcd
+		// backend for those three sub DBs. But we namespace it properly
+		// to make such a split even easier in the future. This will
+		// break lnd for users that ran on etcd with 0.13.x since that
+		// code used the root namespace. We assume that nobody used etcd
+		// for mainnet just yet since that feature was clearly marked as
+		// experimental in 0.13.x.
+		etcdBackend, err := kvdb.Open(
+			kvdb.EtcdBackendName, ctx,
+			db.Etcd.CloneWithSubNamespace(NSChannelDB),
```

Comment:
I also ran into this line: 

https://github.com/lightningnetwork/lnd/blob/f4da59ec1cd960650384578101084ad6557f585d/batch/batch.go#L74

Here the error isn't checked, because `failIdx >= 0` must mean that an error occurred. Only with the new tx retries, this is no longer true? If a serialization error occurred somewhere and the retry succeeded, `failIdx` is still set and the error logic is applied.

Reply:
Has etcd been tested with and without single writer on for example the benchmark to see how much the performance gain actually is? Serializable isolation is not necessarily faster if a lot of collisions happen.

Reply:
Good find! We'll run extensive benchmark and other tests once this PR and some of the other fixes have landed in master.
I've started a new general issue to keep track of all the things that will be found in the process: https://github.com/lightningnetwork/lnd/issues/5557